src/NHibernate/Logging.cs

```diff
@@ -316,118 +738,254 @@ static Log4NetLogger()
 
 		public void ErrorFormat(string format, params object[] args)
 		{
 			if (IsErrorEnabled)
-				ErrorFormatDelegate(logger, format, args);
+				ErrorFormatDelegate(_logger, format, args);
 		}
 
 		public void Fatal(object message)
 		{
 			if (IsFatalEnabled)
-				FatalDelegate(logger, message);
+				FatalDelegate(_logger, message);
 		}
 
 		public void Fatal(object message, Exception exception)
 		{
 			if (IsFatalEnabled)
-				FatalExceptionDelegate(logger, message, exception);
+				FatalExceptionDelegate(_logger, message, exception);
 		}
 
 		public void Debug(object message)
 		{
 			if (IsDebugEnabled)
-				DebugDelegate(logger, message);
+				DebugDelegate(_logger, message);
 		}
 
 		public void Debug(object message, Exception exception)
 		{
 			if (IsDebugEnabled)
-				DebugExceptionDelegate(logger, message, exception);
+				DebugExceptionDelegate(_logger, message, exception);
 		}
 
 		public void DebugFormat(string format, params object[] args)
 		{
 			if (IsDebugEnabled)
-				DebugFormatDelegate(logger, format, args);
+				DebugFormatDelegate(_logger, format, args);
 		}
 
 		public void Info(object message)
 		{
 			if (IsInfoEnabled)
-				InfoDelegate(logger, message);
+				InfoDelegate(_logger, message);
 		}
 
 		public void Info(object message, Exception exception)
 		{
 			if (IsInfoEnabled)
-				InfoExceptionDelegate(logger, message, exception);
+				InfoExceptionDelegate(_logger, message, exception);
 		}
 
 		public void InfoFormat(string format, params object[] args)
 		{
 			if (IsInfoEnabled)
-				InfoFormatDelegate(logger, format, args);
+				InfoFormatDelegate(_logger, format, args);
 		}
 
 		public void Warn(object message)
 		{
 			if (IsWarnEnabled)
-				WarnDelegate(logger, message);
+				WarnDelegate(_logger, message);
 		}
 
 		public void Warn(object message, Exception exception)
 		{
 			if (IsWarnEnabled)
-				WarnExceptionDelegate(logger, message, exception);
+				WarnExceptionDelegate(_logger, message, exception);
 		}
 
 		public void WarnFormat(string format, params object[] args)
 		{
 			if (IsWarnEnabled)
-				WarnFormatDelegate(logger, format, args);
+				WarnFormatDelegate(_logger, format, args);
+		}
+	}
+
+	/// <summary>
+	/// Extensions method for logging.
+	/// </summary>
+	public static class NHibernateLoggerExtensions
+	{
+		public static bool IsDebugEnabled(this INHibernateLogger logger) => logger.IsEnabled(InternalLogLevel.Debug);
+		public static bool IsInfoEnabled(this INHibernateLogger logger) => logger.IsEnabled(InternalLogLevel.Info);
+		public static bool IsWarnEnabled(this INHibernateLogger logger) => logger.IsEnabled(InternalLogLevel.Warn);
+		public static bool IsErrorEnabled(this INHibernateLogger logger) => logger.IsEnabled(InternalLogLevel.Error);
+		public static bool IsFatalEnabled(this INHibernateLogger logger) => logger.IsEnabled(InternalLogLevel.Fatal);
+
+		public static void Fatal(this INHibernateLogger logger, Exception exception, string format, params object[] args)
```

Comment:
`params` always allocates an array. Better to provide several overloads, which take no args, one arg, two args, etc.

Reply:
> `params` always allocates an array. Better to provide several overloads, which take no args, one arg, two args, etc.

An overload with "no args" would save the array, but I do not see how overloads for one args, two args, ..., would save it.
Unless the underlying structure `InternalLogValues` handles additional properties for storing separately individual values of arguments for those other overloads (which would quite complicate the struct and its usage), it would anyway have to create an array for storing them.

So why not adding overload for the "no args" cases, but for the other cases it seems to me quite harder to get benefits from coding them.

By the way, the previous interface is not attempting to optimize these other cases. It is having "no args" methods, but not "one arg", "two args", ... Introducing optimization for these other args cardinality cases may ask for another PR.